KAFKA-9893: Configurable TCP connection timeout and improve the initial metadata fetch

ctan888 opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

1. Added a new common client configuration parameter socket.connection.setup.timeout.ms to the NetworkClient. Handle potential transportation layer timeout using the same approach as it handling potential request timeout.
2. When no connected channel exists, leastLoadedNode() will now provide a disconnected node that has the least recent connection attempts. 
3. ClusterConnectionStates will keep the connecting node ids. Now it also has several new public methods to provide per connection relavant data.

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

1. A unit test for the basic functionality of the new configuration
2. A unit test for optimized leastLoadedNode()
3. Use docker scenarios for system tests. May add a duckertape test.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ctan888 committed:
Adding client configs and change the signature of NetworkClient

ctan888 committed:
Clean the config names and types

ctan888 committed:
Added testLeastLoadedNodeProvideDisconnectedNodesPrioritizedByFailedAttempts(), fixed the failedAttempts counting bug

ctan888 committed:
fix state transition

ctan888 committed:
Migrate static timeout to exponential timeout

ctan888 committed:
Finished the Javadoc, adding more tests

ctan888 committed:
NIT, and apply config to more clients

ctan888 committed:
resolve merge conflict

ctan888 committed:
Fix BrokerApiVersionsCommand

ctan888 committed:
Support connection timeout in KafkaConfig

dajac commented:
@d8tltanc Is this one ready for being reviewed? I can help doing it if it is.

ctan888 commented:
@dajac Hi David. Yes, this patch is ready for review. Thanks.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -34,19 +36,22 @@
-        this.reconnectBackoffMaxMs = reconnectBackoffMaxMs;
-        this.reconnectBackoffMaxExp = Math.log(this.reconnectBackoffMaxMs / (double) Math.max(reconnectBackoffMs, 1)) / Math.log(RECONNECT_BACKOFF_EXP_BASE);
+        this.reconnectBackoff = new GeometricProgression(
+                reconnectBackoffMs, 2, reconnectBackoffMaxMs, 0.2);
```
Just to be sure, we are reusing the code, but not changing the behaviour for backoff right?

ctan888 replied:
Yes, we are reusing the code.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -34,19 +36,22 @@
 final class ClusterConnectionStates {
-    private final long reconnectBackoffInitMs;
-    private final long reconnectBackoffMaxMs;
-    private final static int RECONNECT_BACKOFF_EXP_BASE = 2;
```
why did we remove the constance and hard-code the number inline?

ctan888 replied:
I was thinking that the existing code hardcoded the jitter. I added back the constance and also added a constance for jitter.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -158,9 +165,16 @@ public InetAddress currentAddress(String id) throws UnknownHostException {
+        nodeState.failedAttempts++;
+        nodeState.reconnectBackoffMs = reconnectBackoff.term(nodeState.failedAttempts - 1);
+        if (nodeState.state == ConnectionState.CONNECTING) {
+            nodeState.connectionSetupTimeoutMs = connectionSetupTimeout.term(nodeState.failedAttempts);
```
`failedAttempts` isn't really the number of failed connections, so not sure what we are setting timeout to.

ctan888 replied:
Correct. I'll add a new class property to record connection failures separately.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -296,36 +314,6 @@ public AuthenticationException authenticationException(String id) {
-     *
-     * @param nodeState The node state object to update
-     */
-    private void resetReconnectBackoff(NodeConnectionState nodeState) {
```
Can't we keep this method to perform reset (perhaps rename the method) and include all types of reset?

ctan888 replied:
Including all types of reset together is probably not a good choice because the reset of failed attempts and the reset of the connection timeout may happen in different places. 

However, I agree we should have some abstraction on the update and reset logic. I'll put the logic in new class methods.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,12 @@
         Utils.join(SecurityProtocol.names(), ", ") + ".";
     public static final String DEFAULT_SECURITY_PROTOCOL = "PLAINTEXT";
 
+    public static final String SOCKET_CONNECTIONS_SETUP_TIMEOUT_MS_CONFIG = "socket.connections.setup.timeout.ms";
```
Didn't we decide to call this `socket.connection.setup.timeout.ms` during KIP discussion? (connection instead of connections)

ctan888 replied:
I think in the kip I used `socket.connections.setup.timeout.ms`
Since @cmccabe also suggests `socket.connection.setup.timeout.ms`, I'll change the KIP proposal to stick to `socket.connection.setup.timeout.ms`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -554,6 +571,7 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
         handleDisconnections(responses, updatedNow);
         handleConnections();
         handleInitiateApiVersionRequests(updatedNow);
+        handleTimeOutConnections(responses, updatedNow);
```
`TimeOut` => `TimedOut`

ctan888 replied:
Do you mean TimeOut => Timeout? Refactored.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +345,37 @@ private NodeConnectionState nodeState(String id) {
+     * @param now the current time in ms
+     */
+    public boolean isConnectionSetupTimeout(String id, long now) {
+        return now - lastConnectAttemptMs(id) > connectionSetupTimeoutMs(id);
```
Shouldn't we also check that the node is in connecting state?

ctan888 replied:
I think we don't need to check if the node is in connecting state because the caller is only applying this test to all the nodes in the connecting state.

dajac replied:
That is indeed true today as the caller only calls with nodes in `connectingNodes` but that may not be true forever. I would add the check as suggested by Rajini here to make the implementation safe.

ctan888 replied:
Make sense. Checker added.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -786,6 +808,26 @@ private void handleAbortedSends(List<ClientResponse> responses) {
+        Set<String> connectingNodes = connectionStates.connectingNodes();
+        for (String nodeId: connectingNodes) {
+            if (connectionStates.isConnectionSetupTimeout(nodeId, now)) {
+                // close connection to the node
```
comment unnecessary since it is obvious from the line below

ctan888 replied:
Line deleted.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -786,6 +808,26 @@ private void handleAbortedSends(List<ClientResponse> responses) {
+            if (connectionStates.isConnectionSetupTimeout(nodeId, now)) {
+                // close connection to the node
+                this.selector.close(nodeId);
+                log.debug("Disconnecting from node {} due to socket connection setup timeout.", nodeId);
```
Can we include the timeout in the log line?

ctan888 replied:
Added timeout value to the log.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java:
```diff
@@ -149,6 +155,16 @@
+                                        CommonClientConfigs.SOCKET_CONNECTIONS_SETUP_TIMEOUT_MS_DOC)
+                                .define(SOCKET_CONNECTIONS_SETUP_TIMEOUT_MAX_MS_CONFIG,
+                                        Type.LONG,
+                                        127 * 1000,
```
Can we define the defaults in `CommonClientConfigs`?

ctan888 replied:
I think the current practice is to hard code the defaults in each client's config definition. Do we still want to define the defaults in CommonClientConfigs?

dajac replied:
While I also recognize that we are not consistent with this, I would do it as suggested by Rajini. The defaults are the same everywhere so it makes sense to have it defined centrally for now.

ctan888 replied:
Sounds good. WIll refactor.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/utils/GeometricProgression.java:
```diff
@@ -0,0 +1,51 @@
+ * If scaleFactor is greater or equal than termMax, a constant term of will be provided
+ * This class is thread-safe
+ */
+public class GeometricProgression {
```
This is not exactly a geometric progression?

ctan888 replied:
The naming is kind of hard because exponential reconnect backoff, exponential retry backoff, and exponential timeout are sharing this util class at the same time. Any suggestion?

dajac replied:
I would go with `ExponentialBackoff` even if we use it for computing an exponential timeout as well. I think that people will understand this.

ctan888 replied:
Good idea. Will go for `ExponentialBackoff`

rajinisivaram reviewed (commented):
@d8tltanc Thanks for the PR, looks good. Left some minor comments.

ctan888 committed:
Rename config, extract reset and update logic, some small NIT refactor

ctan888 commented:
Thanks for the feedback @rajinisivaram 
I've modified the patch per your suggestions.
Please take your time and let me know if we can start testing.

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -300,30 +323,48 @@ public AuthenticationException authenticationException(String id) {
-        }
+    private void incrementReconnectBackoff(NodeConnectionState nodeState, long now) {
+        nodeState.reconnectBackoffMs = reconnectBackoff.term(nodeState.failedAttempts);
+        nodeState.failedAttempts++;
```
Shouldn't we increment before computing the new reconnect backoff?

ctan888 replied:
reconnectBackoff.term(0) will return the ${reconnect.backoff.ms} * 2 ^ 0 * jitter
reconnectBackoff.term(1) will return the ${reconnect.backoff.ms} * 2 ^ 1 * jitter

The difference btw reconnect backoff and connection timeout here is that, after the first failed attempts, the connection timeout will be the 1st term of the randomized geometric sequence but the reconnect backoff will be the 0st term of the randomized geometric sequence. So We should use (failedAttempts - 1) for fetching reconnect backoff and (failedAttemps) for fetching the connection timeout.

dajac replied:
That makes sense. Thanks for the clarification.

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +398,36 @@ private NodeConnectionState nodeState(String id) {
+     * @param now the current time in ms
+     */
+    public boolean isConnectionSetupTimeout(String id, long now) {
+        return now - lastConnectAttemptMs(id) > connectionSetupTimeoutMs(id);
```
Are you sure that using `lastConnectAttemptMs` is correct here? `lastConnectAttemptMs` is recorded when a connection is disconnected and as we respect the `reconnectBackoffMs` before reconnecting, the connection timeout computed here will also include the current `reconnectBackoffMs`. Is this what we want? It may be better to record the time in `connecting`.

ctan888 replied:
I think so. The `lastConnectAttemptMs` is updated in both `connecting` (Line 145 & Line 157) and `disconnected`.

dajac replied:
Oh right. I missed the ones in `connecting`. Thanks for the clarification.

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +398,36 @@ private NodeConnectionState nodeState(String id) {
+
+    public long connectionSetupTimeoutMs(String id) {
+        NodeConnectionState nodeState = this.nodeState.get(id);
+        return nodeState.connectionSetupTimeoutMs;
```
Should we ensure that `nodeState` is not `null` here?

ctan888 replied:
No. The caller will ensure that the node is in the connecting state. I'll add an IllegalStateException here.

dajac replied:
I would rather prefer to handle this like we did in `lastConnectAttemptMs` in order to remain consistent. If `nodeState` is `null`, we can return `0`.

ctan888 replied:
When `NetworkClient` initializes a connection to a given node (`NetworkClient::initiateConnect`), it's guaranteed that the `nodeState` will get initialized and won't be `null`. I think it's probably not reasonable if the caller wants to get the connection timeout of a given node before the connection initialization, which is the reason I prevent this kind of calling by throwing the exception.

However, it might be reasonable for a caller to get the `lastConnectAttemptMs` before initializing the connection. For example, the node provider wants to provide a node with the least recent connection attempt. For those nodes haven't been connected yet, their `NodeConnectionState` does not exist. However, this implies that the node has the highest priority and we may assume their `lastConnectAttemptMs` is 0.

dajac replied:
Thanks for the clarification. In this case, let's reuse the `nodeState` method which check null and throws an `IllegalStateException` as you do here. We may be able to use it in `isConnectionSetupTimeout` as well.

ctan888 replied:
Sure. Now `isConnectionSetupTimeout` is also using this checker.

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -678,7 +696,11 @@ public Node leastLoadedNode(long now) {
                 foundConnecting = node;
             } else if (canConnect(node, now)) {
-                foundCanConnect = node;
+                if (foundCanConnect == null ||
+                        this.connectionStates.lastConnectAttemptMs(foundCanConnect.idString()) >
+                                this.connectionStates.lastConnectAttemptMs(node.idString())) {
+                    foundCanConnect = node;
+                }
```
It would be great if you could update the javadoc of the method to reflect this.

ctan888 replied:
Yes

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -786,6 +808,29 @@ private void handleAbortedSends(List<ClientResponse> responses) {
+     */
+    private void handleTimeoutConnections(List<ClientResponse> responses, long now) {
+        Set<String> connectingNodes = connectionStates.connectingNodes();
+        for (String nodeId: connectingNodes) {
```
nit: We usually put a space before and after the `:`.

ctan888 replied:
Refactored

dajac commented on clients/src/test/java/org/apache/kafka/common/utils/GeometricProgressionTest.java:
```diff
@@ -0,0 +1,45 @@
+                assertEquals(scaleFactor * Math.pow(ratio, n), geometricProgression.term(n),
+                        scaleFactor * Math.pow(ratio, n) * jitter);
+            }
+            System.out.println(geometricProgression.term(5));
```
This one can be removed I suppose.

ctan888 replied:
Oh, right. I forgot removing it.

dajac commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,12 @@
     public static final String DEFAULT_SECURITY_PROTOCOL = "PLAINTEXT";
 
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG = "socket.connection.setup.timeout.ms";
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DOC = "The amount of time the client will wait for the initial socket connection to be built. If the connection is not built before the timeout elapses the network client will close the socket channel. The default value will be 10 seconds.";
```
nit: I would use `to be established` instead of `to be built` and I think that we should avoid putting default values in the documentation here because first the default can be changed on a per client basis and second the default will be documented based on the default value provided in config definition.

ctan888 replied:
Make sense. I'll change the description and remove the defaults in the doc.

dajac commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,12 @@
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DOC = "The amount of time the client will wait for the initial socket connection to be built. If the connection is not built before the timeout elapses the network client will close the socket channel. The default value will be 10 seconds.";
+
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG = "socket.connection.setup.timeout.max.ms";
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_DOC = "The maximum amount of time the client will wait for the initial socket connection to be built. The connection setup timeout will increase exponentially for each consecutive connection failure up to this maximum. To avoid connection storms, a randomization factor of 0.2 will be applied to the backoff resulting in a random ...
```
Same comment as above.

ctan888 replied:
Refactored

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/GeometricProgression.java:
```diff
@@ -0,0 +1,51 @@
+                Math.log(termMax / (double) Math.max(scaleFactor, 1)) / Math.log(ratio) : 0;
+    }
+
+    public long term(long n) {
```
nit: What about using `retries` instead of `n`? It may help to understand that the exponential value is computed based on the number of tries or retries.

ctan888 replied:
As we noticed in your earlier comments, the same value of `attempts` may correspond to different terms.

connection_timeout = constant * 2 ^ (attempts)
reconnect_backoff = constant * 2 ^ (attempts - 1)
(in KIP-580) retry_backoff = constant * 2 ^ (attempts - 1) 

So I think using `retries` or `attempts` instead of `n` might also confuse people. Shall we think of another naming?

dajac reviewed (commented):
@d8tltanc Thanks for the PR. Overall, it looks pretty good. I have left some comments. I haven't fully looked at the tests yet.

ctan888 committed:
Second round refactoring

ctan888 commented:
Thanks, @dajac for the comments. I've modified the PR per your suggestions. 
@rajinisivaram Do you think we can start testing?

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +399,38 @@ private NodeConnectionState nodeState(String id) {
+        return nodeState == null ? 0 : nodeState.lastConnectAttemptMs;
+    }
+
+    public long connectionSetupTimeoutMs(String id) {
```
Can we add a javadoc for this method?

ctan888 replied:
Yes. Added.

dajac commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -185,4 +191,9 @@ public static void warnIfDeprecatedDnsLookupValue(AbstractConfig config) {
     }
+
+    public class Defaults {
+        public static final long SocketConnectionSetupTimeoutMs = 10 * 1000;
+        public static final long SocketConnectionSetupTimeoutMaxMs = 127 * 1000;
```
We usually defined constants with capital letters and underscores to separate words. Moreover, we usually put defaults next to the config and the doc. We could use the following:
- `SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DEFAULT`; and
- `SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_DEFAULT`.

rajinisivaram replied:
In KafkaConfigs alone, we separated the defaults. For this one, we should do as @dajac suggested above (similar one is DEFAULT_SECURITY_PROTOCOL).

ctan888 replied:
I see. Refactored.

dajac commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -771,7 +794,7 @@ private void processDisconnection(List<ClientResponse> responses,
      * @param now The current time
      */
-    private void handleTimedOutRequests(List<ClientResponse> responses, long now) {
+    private void handleTimeoutRequests(List<ClientResponse> responses, long now) {
```
nit: I personally prefer the previous name. I would also rename `handleTimeoutConnections` to `handleTimedOutConnections`.

rajinisivaram replied:
I agree, the previous name was better

ctan888 replied:
Sure. Refactored.

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java:
```diff
@@ -0,0 +1,51 @@
+    private final long scaleFactor;
+    private final double jitter;
+
+    public ExponentialBackoff(long scaleFactor, int ratio, long termMax, double jitter) {
```
I find the terminology used here not intuitive. What about using something like: `initialInterval`, `multiplier` and `maxInterval`? I think these are more common when it comes to configuring backoff.

ctan888 replied:
I was using the naming for geometric sequences on Wikipedia. I think your suggestion also make sense. @rajinisivaram Do you think we can use `initialInterval`, `multiplier`, and `maxInterval`?

dajac replied:
@rajinisivaram What do you think about the proposal?

rajinisivaram replied:
Sorry, missed this earlier. Yes, I prefer `initialInterval`, `multiplier`, and `maxInterval` too since they fit better with  ExponentialBackoff.

ctan888 replied:
Sure. Let's use this proposal.

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java:
```diff
@@ -0,0 +1,51 @@
+                Math.log(termMax / (double) Math.max(scaleFactor, 1)) / Math.log(ratio) : 0;
+    }
+
+    public long term(long n) {
```
nit: Let me try with another suggestion ;) What about `backoff(long attempts)`?

rajinisivaram replied:
`term` is non-intuitive. `backoff` doesn't quite fit for connection timeout, but I guess it fits with `ExponentialBackoff`. So unless there is a better suggestion, `backoff` seems reasonable.

ctan888 replied:
Thanks for the suggestion. Let's use the signature `backoff(long attempts)` for now

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * An utility class for exponential backoff, timeout, etc...
```
nit: I would rephrase this a bit and also explain briefly how we use it in AK (e.g. connect timeout, reconnection backoff, etc.). Also, I would suggest to terminate phrases with `.`.

ctan888 replied:
Sure. How does it look like now?

dajac commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -321,4 +325,37 @@ public void testIsPreparingConnection() {
+        assertTrue(connectionStates.canConnect(nodeId1, time.milliseconds()));
+
+        // Check the exponential timeout growth
+        for (int n = 0; n <= Math.log((double) connectionSetupTimeoutMaxMs / connectionSetupTimeoutMs) / Math.log(2); n++) {
```
nit: Could we define a local constant for the base `2` and reuse it everywhere?

ctan888 replied:
Yes. I added static class properties to make the test robust.

dajac commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -321,4 +325,37 @@ public void testIsPreparingConnection() {
+            assertTrue(connectionStates.connectingNodes().contains(nodeId1));
+            assertEquals(connectionSetupTimeoutMs * Math.pow(2, n),
+                    connectionStates.connectionSetupTimeoutMs(nodeId1),
+                    connectionSetupTimeoutMs * Math.pow(2, n) * 0.2);
```
nit: For the jitter, what about defining a constant like `reconnectBackoffJitter`?

ctan888 replied:
Yes. I added static class properties for specifying these parameters.

dajac commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -321,4 +325,37 @@ public void testIsPreparingConnection() {
+
+        // Check the timeout value upper bound
+        connectionStates.connecting(nodeId1, time.milliseconds(), "localhost", ClientDnsLookup.DEFAULT);
+        assertEquals(connectionStates.connectionSetupTimeoutMs(nodeId1), connectionSetupTimeoutMaxMs, connectionSetupTimeoutMaxMs * 0.2);
```
nit: I would put `connectionSetupTimeoutMaxMs` first as it is the expected value. The same applies below.

ctan888 replied:
Good catch. The `expected` should be what you suggested. Refactored.

dajac commented on core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala:
```diff
@@ -205,6 +205,8 @@ object BrokerApiVersionsCommand {
   private object AdminClient {
     val DefaultConnectionMaxIdleMs = 9 * 60 * 1000
     val DefaultRequestTimeoutMs = 5000
+    val DefaultSocketConnectionSetupMs = 10 * 1000
+    val DefaultSocketConnectionSetupMaxMs = 127 * 1000
```
It seems that we can remove these two and use the ones in `CommonClientConfigs`.

ctan888 replied:
Yes. Removed.

rajinisivaram replied:
Looks like the constants haven't been removed.

ctan888 replied:
Oh, that's true. Removed them.

dajac commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -88,6 +88,8 @@ object Defaults {
   val MaxConnections: Int = Int.MaxValue
   val ConnectionsMaxIdleMs = 10 * 60 * 1000L
   val RequestTimeoutMs = 30000
+  val ConnectionSetupTimeoutMs = 10 * 1000L
+  val ConnectionSetupTimeoutMaxMs = 127 * 1000L
```
We can use the ones defined in `CommonClientConfigs` here.

ctan888 replied:
Yes. Refactored.

dajac commented on clients/src/test/java/org/apache/kafka/common/utils/ExponentialBackoffTest.java:
```diff
@@ -0,0 +1,44 @@
+        );
+
+        for (int i = 0; i <= 100; i++) {
+            for (int n = 0; n <= 4; n++) {
+                assertEquals(scaleFactor * Math.pow(ratio, n), exponentialBackoff.term(n),
+                        scaleFactor * Math.pow(ratio, n) * jitter);
+            }
+            assertTrue(exponentialBackoff.term(1000) <= termMax * (1 + jitter));
```
I think that it would be good to test with `n=5` as well to ensure that `termMax` is already applied from it. Perhaps, we could do the following:
```
for (int n = 0; n <= 100; n++) {
  if (n <= 4)
    assertEquals(scaleFactor * Math.pow(ratio, n), exponentialBackoff.term(n),
       scaleFactor * Math.pow(ratio, n) * jitter);
  else
    assertTrue(exponentialBackoff.term(1000) <= termMax * (1 + jitter));
}
```

ctan888 replied:
Thanks for the suggestion. I've adopted this in my test.

dajac reviewed (commented):
@d8tltanc Thanks for the updated PR. I have made a second pass and I have left few more comments. The PR looks good overall.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -158,9 +175,15 @@ public InetAddress currentAddress(String id) throws UnknownHostException {
-        nodeState.state = ConnectionState.DISCONNECTED;
         nodeState.lastConnectAttemptMs = now;
-        updateReconnectBackoff(nodeState);
+        incrementReconnectBackoff(nodeState);
```
the attempt number is incremented, but backoff value is not exactly incremented. I think we should leave it as `updateReconnectBackoff`.

ctan888 replied:
Make sense. Refactored.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -158,9 +175,15 @@ public InetAddress currentAddress(String id) throws UnknownHostException {
-        updateReconnectBackoff(nodeState);
+        incrementReconnectBackoff(nodeState);
+        if (nodeState.state == ConnectionState.CONNECTING) {
+            incrementConnectionSetupTimeout(nodeState);
```
`update` instead of `increment` here too?

ctan888 replied:
Refactored

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -300,30 +325,47 @@ public AuthenticationException authenticationException(String id) {
      * value configured via reconnect.backoff.ms
      *
-     * @param nodeState The node state object to update
+     * @param nodeState nodeState The node state object to update
```
`nodeState` added twice?

ctan888 replied:
Removed.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,12 @@
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DOC = "The amount of time the client will wait for the initial socket connection to be established. If the connection is not built before the timeout elapses the network client will close the socket channel.";
+
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG = "socket.connection.setup.timeout.max.ms";
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_DOC = "The maximum amount of time the client will wait for the initial socket connection to be established. The connection setup timeout will increase exponentially for each consecutive connection failure up to this maximum. To avoid connection storms, a randomization factor of 0.2 will be applied to the backoff resulting in a  ...
```
`initial socket connection ` - not initial?

ctan888 replied:
Yes. Refactored.

rajinisivaram reviewed (commented):
@d8tltanc Thanks for the updates, left a few more comments. It is looking good, I think it should be ready to go once all the outstanding comments from @dajac are addressed.

ctan888 committed:
comment adoption

ctan888 commented:
@dajac @rajinisivaram Thanks for the second round of review. I've addressed your comments and adopted your suggestions. Please let me know if you have more thoughts on this PR.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,14 @@
 
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG = "socket.connection.setup.timeout.ms";
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DOC = "The amount of time the client will wait for the socket connection to be established. If the connection is not built before the timeout elapses, clients will close the socket channel.";
+    public static final Long SOCKET_CONNECTION_SETUP_TIMEOUT_MS_DEFAULT = 10 * 1000L;
```
We should be consistent with the naming above for `DEFAULT_SECURITY_PROTOCOL` and name this `DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT_MS`. We seem to use xxx_DEFAULT in some places, but common configs here and in SslConfigs and SaslConfigs use DEFAULT_xxx, so let's stick to that.

ctan888 replied:
Sure. Refactored.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -103,6 +103,14 @@
+
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG = "socket.connection.setup.timeout.max.ms";
+    public static final String SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_DOC = "The maximum amount of time the client will wait for the socket connection to be established. The connection setup timeout will increase exponentially for each consecutive connection failure up to this maximum. To avoid connection storms, a randomization factor of 0.2 will be applied to the timeout resulting in a random r ...
+    public static final Long SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_DEFAULT = 127 * 1000L;
```
As above, `DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS`

ctan888 replied:
Yes.

rajinisivaram reviewed (approved):
@d8tltanc Thanks for the updates, left a couple of minor comments, apart from that LGTM.

rajinisivaram commented:
retest this please

dajac commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -40,7 +40,12 @@
+    private final static int RECONNECT_BACKOFF_EXP_BASE = 2;
+    private final static double RECONNECT_BACKOFF_JITTER = 0.2;
+    private final static int CONNECTION_SETUP_TIMEOUT_EXP_BASE = 2;
+    private final static double CONNECTION_SETUP_TIMEOUT_JITTER = 0.2;
```
nit: It may be better to keep them as lower case to stay inline with the other constants. Or shall we convert all the others to upper case and to static constant?

rajinisivaram replied:
We can just make the constants in `ClusterConnectionStates` package-private and reuse them here?

ctan888 replied:
Make sense. I've made them package-private and lower-cased the constants.

dajac reviewed (commented):
@d8tltanc Thanks for the update. I have left few minor comments. It looks good overall modulo the unresolved question about the naming in the exponential backoff class.

rajinisivaram reviewed (commented):
@dajac @d8tltanc Left a couple of comments.

dajac commented:
@d8tltanc It seems that many tests failed due to an IllegalStateException. It seems that isConnectionTimeout is called while the connection state is not Connecting. Do you have an idea where it comes from?

ctan888 committed:
3rd round of comment adoption

ctan888 commented:
@rajinisivaram @dajac The test failures are caused by the connection state transition from `CONNECTING` to `CHECKING_API_VERSIONS`, and then to `CONNECTED`, instead of to `CONNECTED` directly. In this case, I should remove the node from the `connecting` HashSet when this transfer happens. I've fixed this issue and update the patch.

Also, I've addressed the latest comments. Please let me know if you have more suggestions and if we can re-run the Jenkins tests.

rajinisivaram commented:
retest this please

rajinisivaram commented:
@d8tltanc Looks like we haven't add the new configs to Connect: https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/3103/

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +402,46 @@ private NodeConnectionState nodeState(String id) {
+     * @param id the connection to fetch the state for
+     */
+    public long connectionSetupTimeoutMs(String id) {
+        NodeConnectionState nodeState = this.nodeState.get(id);
+        if (nodeState == null)
+            throw new IllegalStateException("Connection to node " + id + " hasn't been initialized");
```
nit: This can be replaced by `this.nodeState(id)`.

ctan888 replied:
Good catch. Refactored.

dajac commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -357,6 +402,46 @@ private NodeConnectionState nodeState(String id) {
+     * @param now the current time in ms
+     */
+    public boolean isConnectionSetupTimeout(String id, long now) {
+        NodeConnectionState nodeState = this.nodeState.get(id);
```
nit: You can also use `this.nodeState(id)` here.

ctan888 replied:
Good catch. Refactored.

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java:
```diff
@@ -0,0 +1,53 @@
+ * retry backoff, exponential reconnect backoff, exponential timeout, etc.
+ * The formula is
+ * Backoff(attempts) = random(1 - jitter, 1 + jitter) * initialInterval * multiplier ^ attempts
+ * If scaleFactor is greater or equal than termMax, a constant backoff of will be provided
```
Replace `scaleFactor` and `termMax` with the new terms.

ctan888 replied:
Replaced.

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/ExponentialBackoff.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * An utility class for keeping the parameters and providing the value of exponential
+ * retry backoff, exponential reconnect backoff, exponential timeout, etc.
+ * The formula is
```
nit: `The formula is` -> `The formula is:`

ctan888 replied:
: added

dajac reviewed (commented):
@d8tltanc Thanks for the update. I have left few small comments.

ctan888 committed:
4th comment

ctan888 commented:
@rajinisivaram @dajac I've fixed the tests in the connect repo. It's indeed caused by I didn't add the new configs to connect. Please let me know if we can re-trigger the tests. Thanks.

rajinisivaram commented:
retest this please

ctan888 committed:
merge upstream

ctan888 commented:
@rajinisivaram Looks like we have a merge conflict now. Can we restart the test?

ctan888 commented:
retest this please

dajac reviewed (approved):
Thanks for the PR. LGTM.

rajinisivaram commented:
retest this please

rajinisivaram commented:
retest this please

rajinisivaram commented:
ok to test

rajinisivaram commented:
Streams test failures not related, merging to trunk.

guozhangwang commented:
@d8tltanc @rajinisivaram @dajac We observed some exception thrown from our soak with the following stack trace:

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
	at org.apache.kafka.clients.NetworkClient.handleTimedOutConnections(NetworkClient.java:822)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:574)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:265)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:236)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:215)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:419)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:359)
```

Seems related to this PR, could you take a look?

dajac commented:
Investigating this.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -786,6 +809,29 @@ private void handleAbortedSends(List<ClientResponse> responses) {
+                    "The timeout value is {} ms.",
+                    nodeId,
+                    connectionStates.connectionSetupTimeoutMs(nodeId));
+                processDisconnection(responses, nodeId, now, ChannelState.LOCAL_CLOSE);
```
@guozhangwang Thanks for reporting the exception in this code.
@d8tltanc @dajac This code segment is unsafe, we are removing `node` from `connectingNodes` in p`processDisconnection()` while iterating over the set. We must be missing a test too (or we have a test with only one connection).

cwildman commented:
Noticed this didn't make the 2.6 release. Would it be possible to backport this to older 2.x versions? Or is 2.6.1 coming soon? Dealing with some environments with bad settings for tcp_syn. Thanks.
